src/miner.cpp

```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
     return nNewTime - nOldTime;
 }
 
-CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const CScript& scriptPubKeyIn,bool blockstreamCoreCompatible);
+BlockAssembler::BlockAssembler(const CChainParams &_chainparams) : chainparams(_chainparams)
```

Comment:
Hey Tom, thanks for the careful look-over.  As this PR implies, I'm not attempting to improve the design here; I'm backporting a code redesign (which I agree is quite poor) that core did that led to CPFP with a view to getting CPFP in BU.  I'm happy to come back and clean up the design if that's what @gandrewstone wants, but I think trying to improve the poor state of affairs at the same time as backporting and merging in existing BU changes is too much churn in one go.